### PR TITLE
particle-*: Configure antenna switch

### PR DESCRIPTION
**Pull request by chrysn**
### Contribution description

The particle-{argon,boron,xenon} boards all have a uFL connector and a PCB antenna for 2.4GHz radio, connected by a SKY13351-378LF switch configured by a VCTL[12] pin pair.

This sets the default configuration to use the PCB antenna, as driving an unconnected antenna will deteriorate radio perforance. I don't think that the nRF52 will damage itself driving an unconnected antenna, but doing that it's generally very bad practice. With bad conscience I did a brief test with pins configured the other way, and got a 20dB signal increase compared to that test.

### Testing procedure

I only tested this on a Xenon board, as that's the only of those I have; there, I got about 20dB increases in ping RSSIs compared to switching on the other antenna. (Testing the original state is rather pointless as there's floating pin on capacitors then, and the switch's behavior when not exactly one pin is on is undefined). My test output (top: correct; bottom: wrong antenna enabled):

```
> ping6 fe80::cc30:8ef6:582b:14c6%9
ping6 fe80::cc30:8ef6:582b:14c6%9
12 bytes from fe80::cc30:8ef6:582b:14c6: icmp_seq=0 ttl=64 rssi=-49 dBm time=3.765 ms
12 bytes from fe80::cc30:8ef6:582b:14c6: icmp_seq=1 ttl=64 rssi=-52 dBm time=3.765 ms
12 bytes from fe80::cc30:8ef6:582b:14c6: icmp_seq=2 ttl=64 rssi=-52 dBm time=3.765 ms
```
```
> ping6 fe80::cc30:8ef6:582b:14c6%9
ping6 fe80::cc30:8ef6:582b:14c6%9
12 bytes from fe80::cc30:8ef6:582b:14c6: icmp_seq=0 ttl=64 rssi=-69 dBm time=3.766 ms
12 bytes from fe80::cc30:8ef6:582b:14c6: icmp_seq=1 ttl=64 rssi=-70 dBm time=3.765 ms
12 bytes from fe80::cc30:8ef6:582b:14c6: icmp_seq=2 ttl=64 rssi=-69 dBm time=3.765 ms
```

That was produced from a nrf52840-dongle talking to a particle-xenon with the different antenna configurations, both running the gcoap example with additional modules enabled to give me USB Ethernet and serial.

I did not do the tests on the master branch but on [my branch for DFU uploads on the Xenon](https://github.com/RIOT-OS/RIOT/issues/12320); I built it as I'm PR'ing it here for all three boards, but would ask the maintainers of the boards who have programming equipment to test as well.

### CODE REVIEW

**Comment by chrysn** [comment]
Details I forgot to mention:

* [Data sheeet of the switch chip](http://www.skyworksinc.com/uploads/documents/SKY13351_378LF_201132I.pdf)
* [Xenon schematics](https://docs.particle.io/datasheets/mesh/xenon-datasheet/#nrf52840)
* [Argon schematics](https://docs.particle.io/datasheets/wi-fi/argon-datasheet/#nrf52840)
* [Boron schematics](https://docs.particle.io/datasheets/cellular/boron-datasheet/#nrf52840) – this is slightly different as there's a dedicated inverter to drive one of the antenna selection lines.

I found nothing in the data sheets that describes the antenna selection, as in their environment it's handled by their software (antennas can be switched using a provided function, and the choice is persisted).

File: boards/common/particle-mesh/board.c [hunk 1]
```diff
@@ -33,6 +33,16 @@ void board_init(void)
     gpio_init(LED2_PIN, GPIO_OUT);
     gpio_set(LED2_PIN);
 
+    gpio_init(VCTL1_PIN, VCTLn_MODE);
+    // Suppress output to the UFL connector
```
Comment (aabadie) [inline boards/common/particle-mesh/board.c]: Please use C-style comments!

**Comment by aabadie** [comment]
There's a typo in the commit message: s/perforance/perfor**m**ance/

**Comment by chrysn** [comment]
Comments adapted (sorry, gotta get a code style check into my workflow), commit message will be changed at squashing time.

**Comment by benpicco** [comment]
I think you should also have a config option / define to select the UFL connector when an antenna is connected.

**Comment by chrysn** [comment]
On Tue, Oct 01, 2019 at 06:44:01AM -0700, benpicco wrote:
> I think you should also have a config option / define to select the
> UFL connector when an antenna is connected. 

What should that be? A a define to be set by in the application's
Makefile? An environment variable turned into a define by the
particle-common/Makefile.include? (Either would work, I just don't know
which pattern RIOT follows for "same board but modified").

**Comment by benpicco** [comment]
Good question.
I think the most user-friendly one would be a Makefile variable that gets turned into a define

```Makefile
BOARD_USE_EXT_ANTENNA ?= 0
ifneq (0, $(BOARD_USE_EXT_ANTENNA))
  CFLAGS += -DBOARD_USE_EXT_ANTENNA
endif
```

That would be useful for other boards too!

But I'm not sure about the style / wording or where to best put this, @cladmi maybe you have an idea?

**Comment by cladmi** [comment]
That would just be a `USEMODULE` with a `PSEUDOMODULE` I guess.
Adding different behavior is triggered by a module.

Or for some cases a feature if it requires doing filtering on BSP supporting it or not.

**Comment by chrysn** [comment]
I don't think a module is appropriate as they are boolean (on or off vs variants), additive (no conflicts, just set unions) and generally harmless to pull in (which antenna selection may not be). The BOARD_X ?= approach seems more fitting here, as an application can express both "use the default" and either explicit value.

**Comment by cladmi** [comment]
@leandrolanzieri may no more about this then. Depending if it is "configuration" or "board description".

**Comment by chrysn** [comment]
Pushed a configurability patch atop of it. It uses a new board function for selection of an enum'd antenna, which enables the user to flip the switch at runtime. (Timeout and fallback recommended when switching remotely ;-) ). This causes no overhead in the init function (where the antenna is provided via -D from a make variable with a default), the function gets inlined and removed under normal circumstances.

File: boards/common/particle-mesh/include/board.h [hunk 2]
```diff
@@ -61,6 +61,44 @@ extern "C" {
...
+    BOARD_NRFANTENNA_EXTERNAL,
+};
+
+/** Drive the on-board antenna switch to connect the nRF radio to a given @p
+ * choice of antenna output.
+ * */
+void board_nrfantenna_select(enum board_nrfantenna_selection choice);
+
+#ifdef BOARD_PARTICLE_XENON
+#define VCTL1_PIN           GPIO_PIN(0, 24)
```
Comment (leandrolanzieri) [inline boards/common/particle-mesh/include/board.h]: Maybe some documentation on these pins definitions could be added, and `|| DOXYGEN` to the `#ifdef` so it shows up.

File: boards/common/particle-mesh/Makefile.include [hunk 3]
```diff
@@ -1,3 +1,9 @@
+# This board's nRF radio comes with a uFL antenna plug and a built-in antenna.
```
Comment (leandrolanzieri) [inline boards/common/particle-mesh/Makefile.include]: Maybe some words regarding this antenna selection could be added in the boards documentation.

File: boards/common/particle-mesh/include/board.h [hunk 4]
```diff
@@ -61,6 +61,44 @@ extern "C" {
...
+#ifdef BOARD_PARTICLE_ARGON
+#define VCTL1_PIN           GPIO_PIN(0, 25)
+#define VCTL2_PIN           GPIO_PIN(0, 2)
+#endif
+
+#ifdef BOARD_PARTICLE_BORON
+#define VCTL1_PIN           GPIO_PIN(0, 7)
+#endif
+
+#define VCTLn_MODE          GPIO_OUT
```
Comment (leandrolanzieri) [inline boards/common/particle-mesh/include/board.h]: Is this needed?

Comment (chrysn) [reply boards/common/particle-mesh/include/board.h]: It seems to be customary to define that, it's used in the GPIO initialization, but if that's not a deliberate RIOT pattern but just something that's been sneaking in, I'm happy to remove it. (I'd prefer that.)

Comment (benpicco) [reply boards/common/particle-mesh/include/board.h]: I think this is a case of cargo cult 😉 
What other mode could that pin possibly have?

**Comment by chrysn** [comment]
All current review comments should be addressed by the newly uploaded changes.

**Comment by chrysn** [comment]
… except for the now failing builds ("But I only changed documentation" – one line of non-doc change (enabling doxygen for an ifdef), and got that wrong), fix to be added momentarily.

**Comment by chrysn** [comment]
Both CI steps fail but both complain about the missing squash step – may I squash (into a single commit? Not quite sure what's the project's preferred size of commits…) to fix that?

**Comment by leandrolanzieri** [comment]
> Both CI steps fail but both complain about the missing squash step – may I squash (into a single commit? Not quite sure what's the project's preferred size of commits…) to fix that?

It looks good to me. I think having one commit for the Makefile and board.* changes and another one for the documentation would be ok in this case. If @benpicco and @aabadie don't have further comments you can squash.

I don't have the boards, but I checked the schematics and changes make sense. I trust @chrysn tests on the Xenon board. @aabadie do you have any of the other hardware to try this out?

**Comment by benpicco** [comment]
Please go ahead and squash.

**Comment by chrysn** [comment]
Thanks for the reviews. Squashed into one commit that introduces the basic setting, and one that adds the configurability with all its documentation and API.

**Comment by haukepetersen** [comment]
Perfect, just what I needed :-)

All green and also verified for my Xenon board -> go!

**Pull request merged**